### ISSUE TITLE
Add support for Unix-domain sockets

### ISSUE THREAD

**Issue by fpagliughi** [eclipse-paho/paho.mqtt.c#864]
I'd like to add support for Unix-domain sockets. 

These satisfy the MQTT v5 transport requirement for "ordered, lossless, [bidirectional] stream of bytes", but on a local machine, they provide higher throughput with lower latency and increased security in comparison to TCP.

A great use case for MQTT is providing an internal message bus on a remote device or computer, in which a large percentage of the messages are simply passed around inside the local machine. The use of MQTT for this is particularly helpful in IoT gateways where some subset of the messages are then bridged to a cloud server in a more traditional MQTT usage. But using UNIX sockets for the internal communications would make them faster and safer.

I have long rallied for this feature to be added to Mosquitto based on the performance metrics, to no avail. But Roger finally relented based on the security argument:
> The main reason that I changed my mind on this was the idea of having localhost only listener, where the access control was through the socket permissions. It's much easier to isolate a trusted listener that way.
https://github.com/eclipse/mosquitto/pull/467

I would suggest for the Paho Client(s) we use a URL in the form:
```
unix://path/to/socket
```

whereas if it's an absolute path, use the additional slash:
```
unix:///absolute/path/to/socket
```
And my biggest hope is that this is actually pretty trivial to implement in the client. Once opened, a Unix socket should mostly just behave as a socket.

**Comment by icraggs** [comment]
I thought about this a long time ago, and I seem to remember implementing at least a trial.  I think, one of the reasons that it hasn't been done so far is that it isn't quite as straightforward as one would hope.  I would be happy to be proved wrong in that, but I wouldn't be surprised either way.

When you say "I'd like to add" ... does that mean you are considering a PR? :-)

**Comment by fpagliughi** [comment]
You might have noticed my "once opened" caveat. Sometimes permissions and security issues trip you up to make a connection, but they're usually well behaved once running. But I was throwing this out to see if anyone could think of any other problems.

And I suppose that I am "considering" a PR. But I have a release of the Rust library to push out this week, then the C++ service release, and then a bigger update on the Rust lib to get it up to "modern" Rust standards.

So if anyone beat me to this, I wouldn't be (a) surprised, or (b) offended.

**Comment by pulquero** [comment]
Came across this for the python client https://pypi.org/project/paho-socket/. Might be a stop-gap for some of us while we are waiting on this issue.

**Comment by vjardinMisc** [comment]
Hi,

I don't see any merge request, is anything ongoing? If not, I'll try to submit a AF_UNIX case.

**Comment by icraggs** [comment]
I'm not aware of any current work on this, no.

**Comment by fpagliughi** [comment]
Sorry, I haven't have any time to get started on this. I would still love to see it get done, so if anyone else wants to get to it before I have the time, please do so, and I will be happy to review and test.

**Comment by fpagliughi** [comment]
OK... this took me how many years?!?

But the PR is up.

**Comment by icraggs** [comment]
@fpagliughi I know the feeling Frank!

Oh that's longer than I thought. Almost 5 years ...

**Comment by fpagliughi** [comment]
Yeah, 5 years for such a small amount of code! I had assumed maybe a 25-50% reduction in latency, but a near 20x performance increase in a core use case (for me) is ridiculous. Had I known, I would have done it sooner. Much sooner!

That makes me wonder about the TCP implementation, though. Perhaps a more thorough performance analysis is warranted. We probably need to take another look at `TCP_NODELAY` and things like that. I'll start looking into it.

But that can wait until v1.3.15 or v1.4. We need to get this out ASAP. I'm working on the C++ to go with v1.3.14. Should be ready within the next few days.

**Comment by vjardin** [comment]
@icraggs and @fpagliughi : even if we could find some TCP improvements, the TCP (localhost) overhead will always be much higher than using a Unix stream socket. 
Some few options to improve the TCP case: TCP_NODELAY, TCP_FASTOPEN but it will never be as close as the AF_UNIX/STREAM support.

**Comment by fpagliughi** [comment]
Yes, but AF_UNIX should be around 2x faster than TCP. Not 20x!

**Comment by icraggs** [comment]
@fpagliughi I've merged the two PRs. There were some small interactions with HTTP proxy processing which I've fixed up. My quick check seemed to work. I'd like to have some CI tests, but that will entail (ideally) adding Unix domain socket support to the test broker, which shouldn't be a big deal but not instant.

So if you want to try it out and let me know how it goes, please do.

**Comment by fpagliughi** [comment]
I already have it working with the C++ library. So far just testing it with the sample apps, but everything is looking good.
